src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,134 @@
+"""
+Perform Chavatal-Gomory strengthening of inequalities
+
+We apply the Chvatal-Goromory procedure to constraints that
+satisfy the following criteria:
+1. All variables with nonzero coefficient must be integral
+2. The constraints are not ranged
+3. If the constraint is of the form ax ⩾ b, then every
+   positive coefficient variable must have finite lowerbound and
+   negative coefficient variables must have finite upperbound.
+
+   If the constraint is of the from ax ⩽ b, then every
+   positive coefficient variable must have finite upperbound and
+   negative coefficient variables must have finite lowerbound.
+
+For example, let ax ⩾ b be a constraint where aᵢ ⩾ 0,
+and every non-zeoro coefficient in a corresponds to an
+integer variable. If there exists an s > 0 such that
+⌈aⱼ s⌉ b / ⌈b s⌉ ⩽ aⱼ
+and there is k such that
+⌈aₖ s⌉ b / ⌈b s⌉ < aₖ (*)
+then we will update ax ⩾ b into ⌈a s⌉ x ⩾ ⌈b s⌉.
+
+We find such a number s heuristically, i.e we try different
+values from the set
+{1, t/a_max, t/a_min, (2t-1)/(2a_min)| t = 1,..., r}
+where a_max = sup(|aᵢ|), a_min = inf(|aᵢ|) and r is resolution which
+is set to be 5 by default.
+
+When performing CG strengthening inequality, we round up positive
+coefficients and round down negative ones, so this procedure
+will not reduce any coefficient to 0.
+"""
+
+function cg_strengthening!(ps::PresolveData{T}, resolution::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+    for i in 1:ps.pb0.ncon
+        ps.rowflag[i] || continue
+        nzind = ps.pb0.arows[i].nzind
+        nzval = ps.pb0.arows[i].nzval
+
+        if isfinite(ps.urow[i]) && isfinite(ps.lrow[i]) ||
+            any(ps.var_types[[j for j in nzind if ps.colflag[j]]] .== CONTINUOUS)
+            # skip ranged constraints
+            # and any constraint that involves continuous variables
+            continue
+        elseif isfinite(ps.lrow[i])
+            s, newbound = lowerbound_cg_inequality(nzval, nzind, ps.lrow[i],
+                                    ps.colflag, ps.lcol, ps.ucol, resolution)
+            if s != 0
+                ps.lrow[i] = newbound
+                ps.pb0.arows[i].nzval = [aij > 0 ? ceil(s*aij) : floor(s*aij) for aij in nzval]
+            end
+        elseif isfinite(ps.urow[i])
+            s, newbound = lowerbound_cg_inequality(-nzval, nzind, -ps.urow[i],
+                                    ps.colflag, ps.lcol, ps.ucol, resolution)
+            if s != 0
+                ps.urow[i] = -newbound
+                ps.pb0.arows[i].nzval = [aij > 0 ? ceil(s*aij) : floor(s*aij) for aij in nzval]
+            end
+        end
+    end
+    return nothing
+end
+
+function lowerbound_cg_inequality(nzval::Vector{T}, nzind::Vector{Int},
+    b::T, colflag::Vector{Bool},
+    var_lb::Vector{T}, var_ub::Vector{T},
+    resolution::Int) where {T}
+    # Finding the CG inequality for constraints for the form ax ⩾ b,
```

Comment:
Instead of a comment, write this in a docstring immediately before the function header.